lib/bacnet-asn1.js

```diff
@@ -1724,29 +1720,57 @@ var bacapp_decode_context_data = function(buffer, offset, maxApduLen, propertyTa
   };
 };
 
-var encode_bacnet_character_string = function(buffer, value) {
-  buffer.buffer[buffer.offset++] = baEnum.BacnetCharacterStringEncodings.CHARACTER_UTF8;
-  buffer.offset += buffer.buffer.write(value, buffer.offset, undefined, 'utf8');
+var encode_bacnet_character_string = function(buffer, value, encoding) {
+  encoding = encoding || baEnum.BacnetCharacterStringEncodings.CHARACTER_UTF8;
+  buffer.buffer[buffer.offset++] = encoding;
+  if (encoding === baEnum.BacnetCharacterStringEncodings.CHARACTER_UTF8) {
+    buffer.offset += buffer.buffer.write(value, buffer.offset, undefined, 'utf8');
+  } else {
+    // Transcode JS string to the target encoding using iconv-lite
+    // Remark: NodeJS v7.0.1 adds: buffer.transcode(source, fromEnc, toEnc)
+    var bufEncoded = iconv.encode(value, get_node_encoding(encoding));
+    buffer.offset += bufEncoded.copy(buffer.buffer, buffer.offset);
+  }
 };
 
-var encode_application_character_string = function(buffer, value) {
+var encode_application_character_string = function(buffer, value, encoding) {
   var tmp = {
     buffer: Buffer.alloc(1472),
     offset: 0
   };
-  encode_bacnet_character_string(tmp, value);
+  encode_bacnet_character_string(tmp, value, encoding);
   encodeTag(buffer, baEnum.BacnetApplicationTags.BACNET_APPLICATION_TAG_CHARACTER_STRING, false, tmp.offset);
   tmp.buffer.copy(buffer.buffer, buffer.offset, 0, tmp.offset);
   buffer.offset += tmp.offset;
 };
 
-var encode_context_character_string = module.exports.encode_context_character_string = function(buffer, tagNumber, value) {
+var encode_context_character_string = module.exports.encode_context_character_string = function(buffer, tagNumber, value, encoding) {
   var tmp = {
     buffer: Buffer.alloc(1472),
     offset: 0
   };
-  encode_bacnet_character_string(tmp, value);
+  encode_bacnet_character_string(tmp, value, encoding);
   encodeTag(buffer, tagNumber, true, tmp.offset);
   tmp.buffer.copy(buffer.buffer, buffer.offset, 0, tmp.offset);
   buffer.offset += tmp.offset;
 };
+
+var get_node_encoding = function(encoding, decodingBuffer, decodingOffset) {
+  switch (encoding) {
+    case baEnum.BacnetCharacterStringEncodings.CHARACTER_UTF8:
+      return 'utf8';
+    case baEnum.BacnetCharacterStringEncodings.CHARACTER_UCS2:
+      if ((decodingBuffer[decodingOffset] === 0xFF) && (decodingBuffer[decodingOffset + 1] === 0xFE)) {
+        return 'ucs2';
+      }
+      return; //UCS-2BE
+    case baEnum.BacnetCharacterStringEncodings.CHARACTER_ISO8859_1:
+      return 'latin1';
+    case baEnum.BacnetCharacterStringEncodings.CHARACTER_UCS4:
```

Comment:
Yes, but actually some are not available at all like the japanese ones are not exactly like specified in BACnet standard. However, in Yabe they have the same probkem and just use `shift_JIS` but they claim it untested. So not sure if we should do the same or just say we don't support these?